Multi GPU partition fixes

gswirski opened the pull request:
This simplifies partition calculations, fixes the wrong buffer size for merging partitions, and provides accessor methods for GPU to determine correct partition sizes.

`min_partition_size` is replaced by `hash_rate` - how many base elements are consumed by one hash iteration. This allows us to calculate at which point it makes sense to start a new partition. `min_partition_size = hash_rate / E::EXTENSION_DEGREE`.

Supersedes https://github.com/facebook/winterfell/pull/339

irakliyk commented on air/src/options.rs:
```diff
@@ -181,9 +181,9 @@ impl ProofOptions {
     pub const fn with_partitions(
         mut self,
         num_partitions: usize,
-        min_partition_size: usize,
+        hash_rate: usize,
     ) -> ProofOptions {
```
Let's also update doc comments on line 180 above.

irakliyk commented on air/src/options.rs:
```diff
@@ -351,49 +351,56 @@ impl Deserializable for FieldExtension {
 #[derive(Debug, Clone, Copy, Eq, PartialEq)]
 pub struct PartitionOptions {
     num_partitions: u8,
-    min_partition_size: u8,
+    hash_rate: u8,
 }
```
It may be good to add a brief explanation about how `num_partitions` and `hash_rate` are used to the doc comments for this struct.

irakliyk commented on air/src/options.rs:
```diff
@@ -351,49 +351,56 @@ impl Deserializable for FieldExtension {
+        if self.num_partitions == 1 {
             return num_columns;
         }
+        let min_partition_size = self.hash_rate as usize / E::EXTENSION_DEGREE;
```
Seems like this can under-estimate `min_partition_size` (assuming it is in base elements). For example, for `hash_rate=8` and `EXTENSION_DEGREE = 3`, we'd get 2 here instead of 6.

irakliyk commented on air/src/options.rs:
```diff
@@ -351,49 +351,56 @@ impl Deserializable for FieldExtension {
             return num_columns;
         }
+        let min_partition_size = self.hash_rate as usize / E::EXTENSION_DEGREE;
         let base_elements_per_partition = cmp::max(
             (num_columns * E::EXTENSION_DEGREE).div_ceil(self.num_partitions as usize),
-            self.min_partition_size as usize,
+            min_partition_size,
         );
 
         base_elements_per_partition.div(E::EXTENSION_DEGREE)
```
Would this logic work correctly in all edge cases? For example, if I did my math correctly, for:

- `EXTENSION_DEGREE = 3`
- `num_columns = 7`
- `hash_rate = 8`
- `num_partitions = 2`.

We'd get 3 as the partition size - which wouldn't be enough to cover all 7 columns with 2 partitions.

gswirski replied:
Great catch, thank you! The old logic was overly complicated and wrong. Hopefully it's fixed now.

irakliyk reviewed (commented):
Looks good! Thank you! I left a few questions/comments inline.

gswirski commented:
All comments have been addressed. Partition sizes are expressed in `E` elements, not base elements, I've added docs to clarify. Also, since it's taking so many attempts to get right, I've also added unit tests with the edge cases you've mentioned.

irakliyk reviewed (approved):
All looks good! Thank you!
